File: src/jvm/main/shad/Nedikov_7/first/LockFreeSet.java
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2021 JetBrains s.r.o.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>
 */

package shad.Nedikov_7.first;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

public class LockFreeSet<T extends Comparable<T>> implements ILockFreeSet<T> {
  private SetNode<T> head = new SetNode<>();
  private AtomicInteger listSize = new AtomicInteger(0);

  @Override
  public boolean add(T value) {
    if (value == null) return false;
    SetNode<T> valNode = new SetNode<>(value);
    SetNode<T> node = head;
    listSize.incrementAndGet();
    while (!node.next.compareAndSet(null, valNode)) {
      node = node.next.get();
      if (node.has(value)) {
        listSize.decrementAndGet();
        return false;
      }
    }
    return true;
  }

  @Override
  public boolean remove(T value) {
    if (value == null) return false;
    SetNode<T> node = head.next.get();
    while (node != null) {
      if (node.elem.compareTo(value) == 0) {
        if (node.deleted.compareAndSet(false, true)) {
          return true;
        }
      }
      node = nextNode(node);
    }
    return false;
  }

  @Override
  public boolean contains(T value) {
    if (value == null) return false;
    int n = listSize.get();
    SetNode<T> node = head.next.get();
    for (int i = 0; node != null && i < n; i++) {
      if (node.has(value)) {
        return true;
      }
      node = nextNode(node);
    }
    return false;
  }

  // delete deleted nodes
  private SetNode<T> nextNode(SetNode<T> node) {
    SetNode<T> nodeNext = node.next.get();
    if (nodeNext == null) return null;
    if (!nodeNext.deleted.get()) return nodeNext;
    if (node.next.compareAndSet(nodeNext, nodeNext.next.get())) {
      listSize.decrementAndGet();
    }
    return node.next.get();
  }

  @Override
  public boolean isEmpty() {
    return !iterator().hasNext();
  }

  @Override
  public Iterator<T> iterator() {
    List<SnapNode<T>> prev = snapshot();
    while (true) {
      List<SnapNode<T>> cur = snapshot();
      if (compareSnapshots(prev, cur)) {
        return cleanupSnapshot(prev).iterator();
      }
      prev = cur;
    }
  }

  private List<T> cleanupSnapshot(List<SnapNode<T>> a) {
    List<T> result = new ArrayList<T>();
    for (SnapNode<T> x : a) {
      if (!x.deleted) {
        result.add(x.link.elem);
      }
    }
    return result;
  }

  private boolean compareSnapshots(List<SnapNode<T>> a, List<SnapNode<T>> b) {
    int n = a.size();
    if (n != b.size()) return false;
    for (int i = 0; i < n; i++) {
      if (!a.get(i).eq(b.get(i))) return false;
    }
    return true;
  }

  private List<SnapNode<T>> snapshot() {
    List<SnapNode<T>> result = new ArrayList<SnapNode<T>>();
    SetNode<T> node = head.next.get();
    while (node != null) {
      result.add(node.copyLW());
      node = node.next.get();
    }
    return result;
  }


  private static class SetNode<T extends Comparable<T>> {
    AtomicReference<SetNode<T>> next = new AtomicReference<>(null); // всегда указывает на следующего соседа или на null
    AtomicBoolean deleted = new AtomicBoolean(false); // меняется только с false на true, но не обратно
    final T elem; // null только в голове

    SetNode(T elem) {
      this.elem = elem;
    }

    SetNode() {
      this(null);
    }

    SnapNode<T> copyLW() {
      return new SnapNode<T>(this, deleted.get());
    }

    boolean has(T value) {
      return !deleted.get() && elem.compareTo(value) == 0;
    }
  }


  private static class SnapNode<T extends Comparable<T>> {
    final boolean deleted;
    final SetNode<T> link;

    SnapNode(SetNode<T> link, boolean deleted) {
      this.link = link;
      this.deleted = deleted;
    }

    boolean eq(SnapNode<T> node) {
      return deleted == node.deleted && link == node.link;
    }
  }
}

File: src/jvm/main/shad/Fedotov_5/first/LockFreeSet.kt
package shad.Fedotov_5.first

import java.util.concurrent.atomic.AtomicStampedReference

/**
 * Lock-Free РјРЅРѕР¶РµСЃС‚РІРѕ.
 * @param <T> РўРёРї РєР»СЋС‡РµР№
</T>
 */
class LockFreeSet<T : Comparable<T>> {

    private val head = Node(Node<T>(null, null), null)


    /**
     * РџСЂРѕРІРµСЂРєР° РјРЅРѕР¶РµСЃС‚РІР° РЅР° РїСѓСЃС‚РѕС‚Сѓ
     *
     * РђР»РіРѕСЂРёС‚Рј РґРѕР»Р¶РµРЅ Р±С‹С‚СЊ РєР°Рє РјРёРЅРёРјСѓРј lock-free
     *
     * @return true РµСЃР»Рё РјРЅРѕР¶РµСЃС‚РІРѕ РїСѓСЃС‚Рѕ, РёРЅР°С‡Рµ - false
     */
    val isEmpty: Boolean
        get() {
        return makeSnapshot().isEmpty()
    }


    /**
     * Р”РѕР±Р°РІРёС‚СЊ РєР»СЋС‡ Рє РјРЅРѕР¶РµСЃС‚РІСѓ
     *
     * РђР»РіРѕСЂРёС‚Рј РґРѕР»Р¶РµРЅ Р±С‹С‚СЊ РєР°Рє РјРёРЅРёРјСѓРј lock-free
     *
     * @param value Р·РЅР°С‡РµРЅРёРµ РєР»СЋС‡Р°
     * @return false РµСЃР»Рё value СѓР¶Рµ СЃСѓС‰РµСЃС‚РІСѓРµС‚ РІ РјРЅРѕР¶РµСЃС‚РІРµ, true РµСЃР»Рё СЌР»РµРјРµРЅС‚ Р±С‹Р» РґРѕР±Р°РІР»РµРЅ
     */
    fun add(value: T): Boolean {
        while (true) {
            val (pred, curr) = find(value)
            if (curr.data != null && curr.data == value) {
                return false
            } else {
                val newNode = Node(null, value)
                newNode.next.set(curr, NOT_DELETED)
                val oldVersion = pred.version
                if (pred.next.compareAndSet(curr, newNode, oldVersion * 2, (oldVersion + 1) * 2)) {
                    return true
                }
            }
        }
    }


    /**
     * РЈРґР°Р»РёС‚СЊ РєР»СЋС‡ РёР· РјРЅРѕР¶РµСЃС‚РІР°
     *
     * РђР»РіРѕСЂРёС‚Рј РґРѕР»Р¶РµРЅ Р±С‹С‚СЊ РєР°Рє РјРёРЅРёРјСѓРј lock-free
     *
     * @param value Р·РЅР°С‡РµРЅРёРµ РєР»СЋС‡Р°
     * @return false РµСЃР»Рё РєР»СЋС‡ РЅРµ Р±С‹Р» РЅР°Р№РґРµРЅ, true РµСЃР»Рё РєР»СЋС‡ СѓСЃРїРµС€РЅРѕ СѓРґР°Р»РµРЅ
     */
    fun remove(value: T): Boolean {
        while (true) {
            val (pred, curr) = find(value)
            if (curr.data != value) {
                return false
            } else {
                val succ = curr.next.reference
                val currVersion = curr.version
                val predVersion = pred.version
                if (curr.next.compareAndSet(succ, succ, currVersion * 2, (currVersion + 1) * 2 + 1)) {
                    pred.next.compareAndSet(curr, succ, predVersion * 2, (predVersion + 1) * 2)
                    return true
                }
            }
        }
    }


    /**
     * РџСЂРѕРІРµСЂРєР° РЅР°Р»РёС‡РёСЏ РєР»СЋС‡Р° РІ РјРЅРѕР¶РµСЃС‚РІРµ
     *
     * РђР»РіРѕСЂРёС‚Рј РґРѕР»Р¶РµРЅ Р±С‹С‚СЊ РєР°Рє РјРёРЅРёРјСѓРј wait-free
     *
     * @param value Р·РЅР°С‡РµРЅРёРµ РєР»СЋС‡Р°
     * @return true РµСЃР»Рё СЌР»РµРјРµРЅС‚ СЃРѕРґРµСЂР¶РёС‚СЃСЏ РІ РјРЅРѕР¶РµСЃС‚РІРµ, РёРЅР°С‡Рµ - false
     */
    operator fun contains(value: T): Boolean {
        var curr = head.next.reference
        while (true) {
            val succ = curr.next.reference
            if (curr.next.stamp % 2 == DELETED) {
                if (curr.data == value) {
                    return false
                }
                curr = succ
            } else {
                val data = curr.data
                if (data == value) {
                    return true
                }
                if (data == null || data > value) {
                    return false
                } else {
                    curr = succ
                }
            }
        }
    }

    /**
     * Р’РѕР·РІСЂР°С‰Р°РµС‚ lock-free РёС‚РµСЂР°С‚РѕСЂ РґР»СЏ РјРЅРѕР¶РµСЃС‚РІР°
     *
     * @return РЅРѕРІС‹Р№ СЌРєР·РµРјРїР»СЏСЂ РёС‚РµСЂР°С‚РѕСЂ РґР»СЏ РјРЅРѕР¶РµСЃС‚РІР°
     */
    operator fun iterator(): Iterator<T> {
        return makeSnapshot().listIterator()
    }

    fun print() {
        var curr = head.next.reference
        while (true) {
            val data = curr.data
            if (data == null) {
                print("--\n")
                return
            } else {
                print("--$data")
            }
            curr = curr.next.reference
        }
    }

    private fun find(value: T): Pair<Node<T>, Node<T>> {
        while (true) {
            var pred = head
            var curr = pred.next.reference
            while (true) {
                val succ = curr.next.reference
                if (curr.next.stamp % 2 == DELETED) {
                    val predVersion = pred.version
                    if (!pred.next.compareAndSet(curr, succ, predVersion * 2, predVersion * 2)) {
                        break
                    }
                    curr = succ
                } else {
                    val data = curr.data
                    if (data == null || data >= value) {
                        return Pair(pred, curr)
                    } else {
                        pred = curr
                        curr = succ
                    }
                }
            }

        }
    }

    private fun makeSnapshot(): List<T> {
        var first = unSafeSnapshot()
        while (true) {
            val second = unSafeSnapshot()
            if (first == second) {
                return second.map { it.second }
            } else {
                first = second
            }
        }
    }

    private fun unSafeSnapshot(): List<Pair<Int, T>> {
        val snapshot = mutableListOf<Pair<Int, T>>()
        var curr = head.next.reference
        while (true) {
            val data = curr.data
            if (data == null) {
                return snapshot
            } else if (curr.next.stamp % 2 == NOT_DELETED) {
                snapshot.add(Pair(curr.version, data))
            }
            curr = curr.next.reference
        }
    }


    private class Node<T>(nextNode: Node<T>?, val data: T?) {
        val next: AtomicStampedReference<Node<T>> = AtomicStampedReference<Node<T>>(nextNode, NOT_DELETED)
        val version: Int
            get() = next.stamp / 2
    }

    companion object {
        const val DELETED = 1
        const val NOT_DELETED = 0
    }
}